### ISSUE TITLE
tokio::fs::File::poll_try_clone is confusing: example clarification or potential new future

### ISSUE THREAD

**Issue by KevinMGranger** [tokio-rs/tokio#848]
## Version
0.1.14

## Platform
Darwin Kevins-MacBook-Pro 17.7.0 Darwin Kernel Version 17.7.0: Fri Nov  2 20:43:16 PDT 2018; root:xnu-4570.71.17~1/RELEASE_X86_64 x86_64

## Subcrates
tokio_io

## Description

[The example on File::poll_try_clone](https://docs.rs/tokio/0.1.14/tokio/fs/struct.File.html#method.poll_try_clone) glosses over what you get when you use `and_then`--  you get an `Async<File>` instead of the File directly. A user might conclude you're not supposed to use that directly, leading them down one of two paths:

1. They try using `future::poll_fn`-- but since that needs to move the original file, the clone becomes useless!
2. They try to investigate if you can trust it to always return `Async::Ready`, and find an undocumented `::blocking_io` method that they (well... I) don't understand, and some talk about it at https://github.com/tokio-rs/tokio/issues/386.

I'm not so sure I understand the code for poll_try_clone, but it looks like you _can't_ always trust it to return `Async::Ready`. 

If you _can_ always rely on that and I'm misunderstanding, we should update the example and documentation to reflect that.

If you _can't_, then we can do something like #850.

Here's an example from before I cleaned it up for the PR:
[run it on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d796619f4a9afd8b4428ca44cb9a2574)

```rust
use tokio::fs::File;
use tokio::prelude::*;

fn demonstrate_current_cloning() {
    let create_and_clone_file_from_example = File::create("./tokio_test")
        .and_then(|mut file| file.poll_try_clone())
        .map(|clone| println!("I have a {:?}", clone))
        .map_err(|err| eprintln!("IO error: {}", err));

    tokio::run(create_and_clone_file_from_example);
}

fn demonstrate_proposal() {
    let create_and_clone_file_cloner = File::create("./tokio_test")
        .and_then(|file| {
            file.try_clone()
                .map(|(file, clone)| println!("I have a {:?} and a {:?}", file, clone))
                .map_err(|(_file, err)| {
                    // drop file but you could branch here and keep working if desired
                    // we could also make the ergonomics for this cleaner for when a failed clone
                    // is fatal anyway. Perhaps a separate future or wrapper future that just drops
                    // the file? `try_critical_clone` ?
                    err
                })
        })
        .map_err(|err| eprintln!("IO error: {}", err));

    tokio::run(create_and_clone_file_cloner);
}

/// Clones a file into two files.
/// Should be used only through Future::fuse, otherwise
/// it will panic if polled after returning successfully.
struct Cloner {
    file: Option<File>,
}

impl Future for Cloner {
    type Item = (File, File);
    type Error = (File, tokio::io::Error);

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        self.file
            .as_mut()
            .unwrap()
            .poll_try_clone()
            .map(|async_| async_.map(|cloned_file| (self.file.take().unwrap(), cloned_file)))
            .map_err(|err| (self.file.take().unwrap(), err))
    }
}

// This is just for ergonomic demonstration.
// The method would be added directly to tokio::fs::File.
trait FileExt: Sized {
    fn try_clone(self) -> future::Fuse<Cloner>;
}

impl FileExt for File {
    /// Try to clone this file.
    fn try_clone(self) -> future::Fuse<Cloner> {
        Cloner { file: Some(self) }.fuse()
    }
}

fn main() {
    demonstrate_current_cloning();
    demonstrate_proposal();
}

```

Finally, if we want to support cloning an arbitrary number of times and getting a Vec of files, we could model that as a Stream.

**Comment by LucioFranco** [comment]
So generally in futures based code functions that start with `poll` are not meant to really be used directly in futures they are generally used in manual future implementations. So that is why its not very clear how it is to be used.

So this poll actually just uses blocking which is used to indicate that closure passed will block the current thread, which is the case for `tokio-fs` since it just uses `std::fs`. If you look inside the blocking_io call it just calls `std::fs::File::try_clone` and that is why it can return an error.

**Comment by carllerche** [comment]
This should be fixed by #850
